tools/rewriter/GenCallAsm.cpp

```diff
@@ -280,6 +286,335 @@ static std::string sig_string(const CAbiSignature &sig,
   return ss.str();
 }
 
+static void emit_fn_call(
+    const std::optional<std::string> target_name,
+    WrapperKind kind,
+    AsmWriter &aw,
+    Arch arch) {
+
+  add_comment_line(aw, "Call wrapped function");
+
+  if (!target_name) {
+    // indirect call
+    assert(kind == WrapperKind::IndirectCallsite);
+    if (arch == Arch::X86) {
+      add_asm_line(aw, "call *%r12");
+    } else if (arch == Arch::Aarch64) {
+      add_asm_line(aw, "br x12"); // TODO, which register?
+    }
+  } else {
+    // direct call
+    if (arch == Arch::X86) {
+      add_asm_line(aw, "call "s + target_name.value());
+    } else if (arch == Arch::Aarch64) {
+      add_asm_line(aw, "b "s + target_name.value());
+    }
+  }
+}
+
+static AsmWriter get_asmwriter(bool as_macro) {
+  std::string terminator = {};
+  // Code generated as a macro needs to terminate each line with '\'
+  if (as_macro) {
+    terminator = "\\"s;
+  }
+  return {.ss = {}, .terminator = terminator};
+}
+
+static void emit_save_regs(AsmWriter &aw, uint32_t caller_pkey, uint32_t target_pkey, Arch arch) {
+  if (arch == Arch::X86) {
+    // Save the old frame pointer and set the frame pointer for the call gate
+    add_asm_line(aw, "pushq %rbp");
+    add_asm_line(aw, "movq %rsp, %rbp");
+    // Save registers that are preserved across function calls before switching to
+    // the other compartment's stack. This is on the caller's stack so it's not in
+    // the diagram above.
+    for (auto &r : preserved_registers) {
+      add_asm_line(aw, "pushq %"s + r);
+    }
+  } else if (arch == Arch::Aarch64) {
+    // TODO ARM stack switching
+    llvm::errs() << "TODO register saving not implemented on ARM\n";
+  }
+}
+
+static void emit_intermediate_pkru(AsmWriter &aw, uint32_t caller_pkey, uint32_t target_pkey, bool returning, Arch arch) {
```

Comment:
Having `if (returning)` makes it hard to see at a glance what assembly is generated in each code path. Can we switch the `bool returning` for an argument with a list of the two registers that need to be saved across the wrpku? So the calls to this function would be
```
emit_intermediate_pkru(..., "rcx", "rdx", ...)
...
emit_intermediate_pkru(..., "rax", "rdx", ...)
```
instead of
```
emit_intermediate_pkru(..., false /* returning */, ...)
...
emit_intermediate_pkru(..., true /* returning */, ...)
```

Reply:
Good idea, I've done this.